validation.go

```diff
@@ -0,0 +1,143 @@
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+const MaxUint = ^uint(0)
+
+// ValidatePartialTree uses leafIndices, leaves and proof to calculate the merkle root of the tree and then compares it
+// to expectedRoot.
+func ValidatePartialTree(leafIndices []uint64, leaves, proof [][]byte, expectedRoot []byte,
+	hash func(lChild, rChild []byte) []byte) (bool, error) {
+	v, err := newValidator(leafIndices, leaves, proof, hash)
+	if err != nil {
+		return false, err
+	}
+	root, err := v.calcRoot(MaxUint)
+	return bytes.Equal(root, expectedRoot), err
+}
+
+func newValidator(leafIndices []uint64, leaves, proof [][]byte,
+	hash func(lChild, rChild []byte) []byte) (validator, error) {
+	if len(leafIndices) != len(leaves) {
+		return validator{}, fmt.Errorf("number of leaves (%d) must equal number of indices (%d)", len(leaves),
+			len(leafIndices))
+	}
+	if len(leaves) == 0 {
+		return validator{}, fmt.Errorf("at least one leaf is required for validation")
+	}
+	proofNodes := &proofIterator{proof}
+	leafIt := &leafIterator{leafIndices, leaves}
+
+	return validator{leaves: leafIt, proofNodes: proofNodes, hash: hash}, nil
+}
+
+type validator struct {
+	leaves     *leafIterator
+	proofNodes *proofIterator
+	hash       func(lChild, rChild []byte) []byte
+}
+
+func (v *validator) calcRoot(stopAtLayer uint) ([]byte, error) {
+	idx, activeNode, err := v.leaves.next()
+	if err != nil {
+		return nil, err
+	}
+	p := position{index: idx}
+	var lChild, rChild, sibling []byte
+	for {
+		if p.height == stopAtLayer {
+			break
+		}
+		nextLeaf, _, err := v.leaves.peek()
+		if err == nil && p.sibling().isAncestorOf(nextLeaf) {
```

Comment:
It will be good to add a comment here explaining that a node which is an ancestor of a leaf can be calculated while the opposite case should be available in the proof

Reply:
Also, I'm afraid that there is some problem with assuming an order on the proof nodes, let's talk about it in person